File: frontend/src/app/services/admin/admin.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { UserBanDTO, BanStatusDTO, UserReportDTO, SkillReportDTO, ReportStatus } from '../../dtos/admin';

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private apiUrl = `${environment.apiBaseUrl}/admin`;

  constructor(private http: HttpClient) { }

  /**
   * Ban a user permanently.
   * @param username Username of the user to ban
   * @param reason Reason for the ban
   * @returns Observable of the created ban
   */
  banUser(username: string, reason: string): Observable<UserBanDTO> {
    return this.http.post<UserBanDTO>(`${this.apiUrl}/ban/${username}`, null, {
      params: { reason }
    });
  }

  /**
   * Check if a user is banned.
   * @param username Username of the user to check
   * @returns Observable of the ban status
   */
  isUserBanned(username: string): Observable<BanStatusDTO> {
    return this.http.get<BanStatusDTO>(`${this.apiUrl}/ban/${username}`);
  }

  /**
   * Report a user for inappropriate behavior.
   * @param reportedUserUsername Username of the user being reported
   * @param reason Reason for the report
   * @returns Observable of the created report
   */
  reportUser(reportedUserUsername: string, reason: string): Observable<UserReportDTO> {
    return this.http.post<UserReportDTO>(`${this.apiUrl}/reports/users/${reportedUserUsername}`, null, {
      params: { reason }
    });
  }

  /**
   * Get all reports for a specific user.
   * @param username Username of the user to get reports for
   * @returns Observable of user reports
   */
  getUserReports(username: string): Observable<UserReportDTO[]> {
    return this.http.get<UserReportDTO[]>(`${this.apiUrl}/reports/users/${username}`);
  }

  /**
   * Get all pending user reports.
   * @returns Observable of pending user reports
   */
  getPendingUserReports(): Observable<UserReportDTO[]> {
    return this.http.get<UserReportDTO[]>(`${this.apiUrl}/reports/users/pending`);
  }

  /**
   * Get all user reports.
   * @returns Observable of all user reports
   */
  getAllUserReports(): Observable<UserReportDTO[]> {
    return this.http.get<UserReportDTO[]>(`${this.apiUrl}/reports/users`);
  }

  /**
   * Report a skill for inappropriate content.
   * @param skillId ID of the skill being reported
   * @param reason Reason for the report
   * @returns Observable of the created report
   */
  reportSkill(skillId: number, reason: string): Observable<SkillReportDTO> {
    return this.http.post<SkillReportDTO>(`${this.apiUrl}/reports/skills/${skillId}`, null, {
      params: { reason }
    });
  }

  /**
   * Get all reports for a specific skill.
   * @param skillId ID of the skill to get reports for
   * @returns Observable of skill reports
   */
  getSkillReports(skillId: number): Observable<SkillReportDTO[]> {
    return this.http.get<SkillReportDTO[]>(`${this.apiUrl}/reports/skills/${skillId}`);
  }

  /**
   * Get all pending skill reports.
   * @returns Observable of pending skill reports
   */
  getPendingSkillReports(): Observable<SkillReportDTO[]> {
    return this.http.get<SkillReportDTO[]>(`${this.apiUrl}/reports/skills/pending`);
  }

  /**
   * Get all skill reports.
   * @returns Observable of all skill reports
   */
  getAllSkillReports(): Observable<SkillReportDTO[]> {
    return this.http.get<SkillReportDTO[]>(`${this.apiUrl}/reports/skills`);
  }

  /**
   * Update the status of a skill report.
   * @param reportId ID of the report to update
   * @param newStatus New status to set
   * @returns Observable of the updated report
   */
  updateSkillReportStatus(reportId: number, newStatus: ReportStatus): Observable<SkillReportDTO> {
    return this.http.put<SkillReportDTO>(`${this.apiUrl}/reports/skills/${reportId}/status`, null, {
      params: { status: newStatus }
    });
  }


  /**
   * Delete a user report.
   * @param reportId ID of the report to delete
   * @returns Observable of void
   */
  deleteUserReport(reportId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/reports/users/${reportId}`);
  }
}

File: backend/src/main/java/at/ac/ase/inso/group02/admin/AdminService.java
package at.ac.ase.inso.group02.admin;

import java.util.List;

import at.ac.ase.inso.group02.admin.dtos.BanStatusDTO;
import at.ac.ase.inso.group02.admin.dtos.SkillReportDTO;
import at.ac.ase.inso.group02.admin.dtos.UserBanDTO;
import at.ac.ase.inso.group02.admin.dtos.UserReportDTO;
import at.ac.ase.inso.group02.entities.admin.ReportStatus;
import jakarta.ws.rs.NotFoundException;

/**
 * Service interface for handling administrative actions such as bans and reports.
 * This service provides methods for managing user bans, user reports, and skill reports.
 */
public interface AdminService {
    /**
     * Permanently bans a user from the platform.
     *
     * @param username The Username of the user to ban
     * @param reason The reason for the ban
     * @return The created UserBanDTO
     * @throws NotFoundException if the user doesn't exist
     * @throws IllegalStateException if the user is already banned
     */
    UserBanDTO banUser(String username, String reason);

    /**
     * Creates a new report against a user.
     *
     * @param reportedUserUsername The Username of the user being reported
     * @param reason The reason for the report
     * @return The created UserReportDTO
     * @throws NotFoundException if either user doesn't exist
     */
    UserReportDTO reportUser(String reportedUserUsername, String reason);

    /**
     * Creates a new report against a skill.
     *
     * @param skillId The ID of the skill being reported
     * @param reason The reason for the report
     * @return The created SkillReportDTO
     * @throws NotFoundException if the skill or user doesn't exist
     */
    SkillReportDTO reportSkill(Long skillId, String reason);

    /**
     * Updates the status of a skill report.
     *
     * @param reportId The ID of the report to update
     * @param newStatus The new status to set
     * @return The updated SkillReportDTO
     * @throws NotFoundException if the report doesn't exist
     */
    SkillReportDTO updateSkillReportStatus(Long reportId, ReportStatus newStatus);

    /**
     * Checks if a user is currently banned.
     *
     * @param username The Username of the user to check
     * @return true if the user is banned, false otherwise
     */
    BanStatusDTO isUserBanned(String username);

    /**
     * Retrieves all reports filed against a specific user.
     *
     * @param username The Username of the user
     * @return List of reports for the user
     */
    List<UserReportDTO> getUserReports(String username);

    /**
     * Retrieves all reports filed against a specific skill.
     *
     * @param skillId The ID of the skill
     * @return List of reports for the skill
     */
    List<SkillReportDTO> getSkillReports(Long skillId);

    /**
     * Retrieves all pending skill reports.
     *
     * @return List of pending skill reports
     */
    List<SkillReportDTO> getPendingSkillReports();

    /**
     * Retrieves all pending user reports.
     *
     * @return List of pending user reports
     */
    List<UserReportDTO> getPendingUserReports();

    /**
     * Retrieves all user reports.
     *
     * @return List of all user reports
     */
    List<UserReportDTO> getAllUserReports();

    /**
     * Retrieves all skill reports.
     *
     * @return List of all skill reports
     */
    List<SkillReportDTO> getAllSkillReports();

    /**
     * Deletes a user report.
     *
     * @param reportId The ID of the report to delete
     * @throws NotFoundException if the report doesn't exist
     */
    void deleteUserReport(Long reportId);
}
